Source/WebCore/page/scrolling/ScrollAnchoringController.cpp

```diff
@@ -25,30 +25,68 @@
 
 #include "config.h"
 #include "ScrollAnchoringController.h"
+#include <wtf/text/TextStream.h>
 
 namespace WebCore {
 
-void ScrollAnchoringController::invalidateAnchorElement()
+void ScrollAnchoringController::invalidateAnchorBox()
 {
-    m_anchorElement = nullptr;
-    m_lastPositionForAnchorElement = FloatPoint();
+    m_anchorBox = nullptr;
+    m_lastPositionForAnchorBox = FloatSize();
 }
 
-void ScrollAnchoringController::selectAnchorElement()
+FloatSize ScrollAnchoringController::computeOffset(RenderBox* candidate)
 {
-    // TODO: implement
-    // perform Anchor Element Selection algorithm on dom elements within frame view's viewable region
-    // Firefox appears to only iterate through frame tree under the current frame (unclear if
-    // looking through frame tree is sufficient)
-    // unclear to me where to call this from 
+    return candidate->location() - m_frameView.rectForFixedPositionLayout().location();
+}
+
+void ScrollAnchoringController::selectAnchorBox()
+{
+    // TODO: fully implement anchor node selection algorithm in spec
+    // (specifically excluded subtrees and priority nodes)
+    auto scrollPosition = frameView().scrollPosition();
+    if (m_anchorBox || scrollPosition == IntPoint())
+        return;
+    
+    auto& document = *m_frameView.frame().document();
+    auto* documentElement = document.documentElement();
+    RenderBox* rootRenderer = documentElement ? documentElement->renderBox() : nullptr;
+        RenderBox* candidateAnchor = nullptr;
+    auto viewableRect = m_frameView.rectForFixedPositionLayout();
+    
+    Vector<RenderBox*> tree;
+    tree.append(rootRenderer->firstChildBox());
+    while (!tree.isEmpty()) {
+        auto box = tree.last();
+        tree.removeLast();
+        if (box->style().overflowAnchor() == OverflowAnchor::None)
+            continue;
+        auto boxRect = box->frameRect();
+        if (viewableRect.intersects(boxRect))
+            candidateAnchor = box;
+        if (viewableRect.contains(boxRect))
+            break;
+        if (!is<RenderBox>(box->firstChild()))
+            continue;
+        for (auto child = box->firstChildBox(); child; child = child->nextSiblingBox()) {
```

Comment:
The spec talks in terms of DOM nodes, not boxes, so does this algorithm need to run on Nodes/Elements?

Reply:
The spec does make mention of boxes: 

> An anchor node can be any [box](https://drafts.csswg.org/css-display-3/#box) except one for a non-[atomic inline](https://drafts.csswg.org/css-display-3/#atomic-inline)
I believe the Google implementation also runs their selection algorithm on the box tree. Functionally it should be the same right?

Reply:
The box tree can differ from the DOM tree because of `display:contents`, `display:none`, fragmentation and many other reasons. Some of these will be observable (e.g. maybe think about scroll anchoring where the anchorBox is something split across 2 columns in multicol).

Reply:
Changed to DOM tree walk.